backend/zeno_backend/server.py

```diff
@@ -468,97 +445,64 @@ def get_project_uuid(owner_name: str, project_name: str, request: Request):
-        return select.projects(user)
-
-    @api_app.get(
-        "/reports-details",
-        response_model=list[ReportDetails],
-        tags=["zeno"],
-    )
-    def get_reports_details(current_user=Depends(auth.claim())):
-        user = select.user(current_user["username"])
-        if user is None:
-            return Response(status_code=status.HTTP_401_UNAUTHORIZED)
-        reports = select.reports(user)
-        report_stats = []
-        for rep in reports:
-            report_stats.append(select.report_stats(rep.id, user.id))
-        return [
-            ReportDetails(report=rep, statistics=report_stats[i])
-            for i, rep in enumerate(reports)
-        ]
-
-    @api_app.get(
-        "/public-reports-details",
-        response_model=list[ReportDetails],
-        tags=["zeno"],
-    )
-    def get_public_reports_details(req: Request):
+    def get_projects(projects_request: ProjectsRequest, req: Request):
         user = util.get_user_from_token(req)
-        reports = select.public_reports()
-        report_stats = []
-        for rep in reports:
-            report_stats.append(select.report_stats(rep.id, user.id if user else None))
-        return [
-            ReportDetails(report=rep, statistics=report_stats[i])
-            for i, rep in enumerate(reports)
-        ]
+        if projects_request.user is None:
+            return select.public_projects(projects_request)
 
-    @api_app.get(
+        if user is None or projects_request.user != user.name:
+            raise HTTPException(
+                status_code=status.HTTP_401_UNAUTHORIZED,
+            )
+        return select.projects(user, projects_request)
+
+    @api_app.post(
         "/reports",
         response_model=list[Report],
         tags=["zeno"],
     )
-    def get_reports(current_user=Depends(auth.claim())):
-        user = select.user(current_user["username"])
-        if user is None:
-            return Response(status_code=status.HTTP_401_UNAUTHORIZED)
-        return select.reports(user)
+    def get_reports(reports_request: ReportsRequest, req: Request):
+        user = util.get_user_from_token(req)
+        if reports_request.user is None:
+            return select.public_reports(reports_request)
+
+        if user is None or reports_request.user != user.name:
+            raise HTTPException(
+                status_code=status.HTTP_401_UNAUTHORIZED,
+            )
+        return select.reports(user, reports_request)
+
+    @api_app.get("/homepage/", response_model=HomepageData, tags=["zeno"])
+    def get_public_homepage_data(req: Request):
+        projects = select.public_projects(ProjectsRequest(limit=6))
+        reports = select.public_reports(ReportsRequest(limit=6))
+        return HomepageData(
+            projects=projects,
+            reports=reports,
+            num_projects=select.project_count(),
+            num_reports=select.report_count(),
+        )
```

Comment:
As said, I would not do this. I don't think we should make our routes match pages but rather objects in the database.

Reply:
I think it's a good bit faster to have a combined request:
https://softwareengineering.stackexchange.com/questions/341872/rest-api-design-multiple-calls-vs-single-call-to-the-api

But not really sure what the right answer is.